### PR TITLE
Persistent collections updates (part 5)

### PR DESCRIPTION
**Pull request by lorentey**
- Reverse the ordering of items in node storage, fixing copying colliding items in iteration order
- Indices now represent paths into the hash tree instead of the preorder position of an item. This speeds up looking up indices (and the item addressed by an index), but it slows down `index(after:)` and `index(before:)`.
- Add index validation logic.
- Add a way to dump trees in preorder walk order
- Implement in-place mutations for the key subscript and the defaulted subscript. (This was a pain in the neck to implement.)
- Add many more tests.
- Fix a logic oversight in the node sizing logic that lead to nodes unnecessarily doubling their size with each new generation of copies.
- Use the term "slot" rather than "offset" for positions within internal storage.

### Checklist
- [X] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-collections)
- [X] My contributions are licensed under the [Swift license](/LICENSE.txt).
- [X] I've followed the coding style of the rest of the project.
- [X] I've added tests covering all new code paths my change adds to the project (if appropriate).
- [ ] I've added benchmarks covering new functionality (if appropriate).
- [X] I've verified that my change does not break any existing tests or introduce unexplained benchmark regressions.
- [ ] I've updated the documentation if necessary.

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
Iteration performance is boosted when the dictionary starts to go beyond a single layer of nodes. 
(I suspect there is some constant overhead here that might be possible to eliminate. Retain/release traffic?)

01 PersistentDictionary Int, Int  indexing subscript
02 PersistentDictionary Int, Int Keys sequential iteration
03 PersistentDictionary Int, Int Values sequential iteration
04 PersistentDictionary Int, Int  sequential iteration, indices

**Comment by lorentey** [comment]
`index(forKey:)` performance is slower than I expect, reinforcing my suspicion about there being some bogus overhead. Looking at generated code will probably make it obvious where the time goes.

05 PersistentDictionary Int, Int  unsuccessful index(forKey:)
26 PersistentDictionary Int, Int  successful index(forKey:)

**Comment by lorentey** [comment]
The new in-place mutation support does show up as a bit of improvement on these microbenchmarks, but support for this is mostly exciting because it prevents CoW copies for the mutated value -- it's not necessary about raw lookup performance.

That said, in-place mutation for the defaulted subscript gets a solid 53% performance boost:

06 PersistentDictionary Int, Int  defaulted subscript, _modify existing
07 PersistentDictionary Int, Int  subscript, _modify
09 PersistentDictionary Int, Int  defaulted subscript, _modify missing

**Comment by lorentey** [comment]
Benchmark summary (compared to #177):

```
Tasks with difference scores larger than 1.05:
  Score   Sum     Improvements Regressions  Name
  17.51   17.51   17.51(#75)   0.9993(#1)   PersistentDictionary<Int, Int> indexing subscript (*)
  1.887   1.887   1.967(#55)   0.9207(#21)  PersistentDictionary<Int, Int>.Keys sequential iteration (*)
  1.881   1.881   1.963(#55)   0.9184(#21)  PersistentDictionary<Int, Int>.Values sequential iteration (*)
  1.738   1.738   1.870(#55)   0.8676(#21)  PersistentDictionary<Int, Int> sequential iteration, indices (*)
  1.596   0.626   1.001(#2)    0.6260(#74)  PersistentDictionary<Int, Int> unsuccessful index(forKey:) (*)
  1.532   1.532   1.532(#75)   0.9997(#1)   PersistentDictionary<Int, Int> defaulted subscript, _modify existing (*)
  1.108   1.108   1.113(#74)   0.9953(#2)   PersistentDictionary<Int, Int> subscript, _modify (*)
  1.086   1.086   1.099(#59)   0.9874(#17)  PersistentDictionary<Int, Int> random removals (missing keys) (*)
  1.081   1.081   1.082(#75)   0.9989(#1)   PersistentDictionary<Int, Int> defaulted subscript, _modify missing (*)
  1.065   1.065   1.069(#60)   0.9964(#16)  PersistentDictionary<Int, Int> subscript, remove existing, unique (*)
  1.064   0.940   1.010(#6)    0.9301(#70)  PersistentDictionary<Int, Int> updateValue(_:forKey:), insert (*)
  1.063   1.063   1.065(#70)   0.9981(#6)   PersistentDictionary<Int, Int> subscript, successful lookups (*)
  1.059   1.059   1.065(#65)   0.9943(#11)  PersistentDictionary<Int, Int> defaulted subscript, successful lookups (*)
```

**Comment by lorentey** [comment]
Last commit fixes `index(forKey:)` performance:

02 PersistentDictionary Int, Int  successful index(forKey:)
07 PersistentDictionary Int, Int  unsuccessful index(forKey:)


```
Tasks with difference scores larger than 1.05:
  Score   Sum     Improvements Regressions  Name
  17.33   17.33   17.33(#75)   0.9993(#1)   PersistentDictionary<Int, Int> indexing subscript (*)
  1.983   1.983   1.983(#76)   1.000(#0)    PersistentDictionary<Int, Int> successful index(forKey:) (*)
  1.867   1.867   1.949(#55)   0.9175(#21)  PersistentDictionary<Int, Int>.Keys sequential iteration (*)
  1.851   1.851   1.937(#55)   0.9142(#21)  PersistentDictionary<Int, Int>.Values sequential iteration (*)
  1.713   1.713   1.843(#55)   0.8699(#21)  PersistentDictionary<Int, Int> sequential iteration, indices (*)
  1.526   1.526   1.526(#76)   1.000(#0)    PersistentDictionary<Int, Int> defaulted subscript, _modify existing (*)
  1.272   1.272   1.272(#76)   1.000(#0)    PersistentDictionary<Int, Int> unsuccessful index(forKey:) (*)
  1.223   1.223   1.223(#76)   1.000(#0)    PersistentDictionary<Int, Int> random removals (missing keys) (*)
  1.184   1.184   1.184(#75)   1.000(#1)    PersistentDictionary<Int, Int> subscript, _modify (*)
  1.183   1.183   1.183(#76)   1.000(#0)    PersistentDictionary<Int, Int> subscript, noop setter (*)
  1.167   1.167   1.167(#76)   1.000(#0)    PersistentDictionary<Int, Int> subscript, remove missing (*)
  1.103   1.103   1.105(#64)   0.9983(#12)  PersistentDictionary<Int, Int> subscript, remove existing, unique (*)
  1.071   1.071   1.076(#66)   0.9957(#10)  PersistentDictionary<Int, Int> defaulted subscript, successful lookups (*)
  1.069   1.069   1.071(#71)   0.9978(#5)   PersistentDictionary<Int, Int> subscript, successful lookups (*)
  1.068   0.936   1.005(#10)   0.9312(#66)  PersistentDictionary<Int, Int> updateValue(_:forKey:), insert (*)
  1.065   1.065   1.067(#65)   0.9987(#11)  PersistentDictionary<Int, Int> random removals (existing keys) (*)
```

**Comment by lorentey** [comment]
@swift-ci test

File: Sources/PersistentCollections/Node/_Node+Mutations.swift [hunk 1]
```diff
@@ -12,13 +12,19 @@
...
   ///
   /// `itemMap` must not already reflect the insertion at the time this
   /// function is called. This method does not update `itemMap`.
+  ///
+  /// - Returns: an unsafe mutable pointer to uninitialized memory that is
+  ///    ready to store the new item. It is the caller's responsibility to
+  ///    initialize this memory.
   @inlinable
-  internal func _insertItem(_ item: __owned Element, at slot: _Slot) {
+  internal func _makeRoomForNewItem(
```
Comment (msteindorfer) [inline Sources/PersistentCollections/Node/_Node+Mutations.swift:25]: 👍

**Review by msteindorfer — APPROVED** [review]
LGTM

File: Sources/PersistentCollections/Node/_Node+Mutations.swift (hunk 1 above)
Comment (lorentey) [reply Sources/PersistentCollections/Node/_Node+Mutations.swift:25]: This was such a pain to plumb through. Ideally we wouldn't need to split the action of restructuring the tree to make room for an item out from the final act of initializing storage, but it's difficult to implement `_modify` accessors well without doing something like this. 

Very soon after this function I switched to delegating initialization/deinitialization of inserted/removed items to closures that get passed in as arguments, which seems to be less painful overall. (It's still quite annoying though.)

File: Sources/PersistentCollections/Node/_Node.swift [hunk 2]
```diff
@@ -38,10 +38,12 @@ internal struct _RawNode {
 /// The storage is arranged as shown below.
 ///
 ///     ┌───┬───┬───┬───┬───┬──────────────────┬───┬───┬───┬───┐
-///     │ 0 │ 1 │ 2 │ 3 │ 4 │→   free space   ←│ 0 │ 1 │ 2 │ 3 │
+///     │ 0 │ 1 │ 2 │ 3 │ 4 │→   free space   ←│ 3 │ 2 │ 1 │ 0 │
 ///     └───┴───┴───┴───┴───┴──────────────────┴───┴───┴───┴───┘
 ///      children                                         items
 ///
```
[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/PersistentCollections/Node/_Node.swift:28]: In order for the empty singleton to work, it's convenient to have a superclass (`_RawStorage`) that is independent the `Key` and `Value` types -- so I couldn't just set `ManagedBuffer`'s `Element` type to `(key: Key, value: Value)`.

Setting it to `_RawNode` instead allows us to rely on `ManagedBuffer` to properly align the start of the storage. (`_RawNode` is layout-compatible with the real, generic `_Node` struct, so I can treat child storage interchangeably as one or the other, depending on the context.)

On the other end of the buffer, we cannot trust that the buffer's end falls on a properly aligned address for an item, so have to make sure we over-allocate slightly so we can round things down to the nearest aligned address. (That code is quite unpleasant, as it's juggling alignment details at the same time as trying to implement a node sizing strategy.)

I could've ended up with some placeholder Element type such as `UInt8` (we end up explicitly rebinding memory anyway), but then I would've had to do manual alignment on both ends, and that felt a bit too much.

File: Sources/PersistentCollections/Node/_Node+UnsafeHandle.swift [hunk 3]
```diff
@@ -171,15 +171,15 @@ extension _Node.UnsafeHandle {
   }
 
   @inlinable
-  internal var _items: UnsafeMutableBufferPointer<Element> {
+  internal var reverseItems: UnsafeMutableBufferPointer<Element> {
```
[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/PersistentCollections/Node/_Node+UnsafeHandle.swift:230]: oh, I accidentally left in the underscore -- I added it as part of a code audit when I reversed the ordering. I'll get rid of it!

File: Sources/PersistentCollections/Node/_Node.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (msteindorfer) [reply Sources/PersistentCollections/Node/_Node.swift:28]: Makes sense, thanks.

**Pull request merged**